/* { dg-options "-Wc++-compat" } */

int bool;
int catch;
int char16_t;
int char32_t;
int class;
int const_cast;			/* { dg-warning "keyword" } */
int decltype;
int delete;
int dynamic_cast;		/* { dg-warning "keyword" } */
int explicit;
int export;
int false;
int friend;
int mutable;			/* { dg-warning "keyword" } */
int namespace;
int new;
int operator;
int private;
int protected;
int public;
int reinterpret_cast;		/* { dg-warning "keyword" } */
int static_assert;		/* { dg-warning "keyword" } */
int static_cast;		/* { dg-warning "keyword" } */
int template;
int this;
int throw;
int true;
int try;
int typename;
int typeid;
int using;
int virtual;
int wchar_t;
