// PR c++/13269
// { dg-options "-fshow-column" }

class Foo { int foo() return 0; } };

// { dg-error "30: error: expected identifier before numeric constant" "" { target *-*-* } { 4 } }

// { dg-error "23: error: named return values are no longer supported" "" { target *-*-* } { 4 } }

// the column number info of this error output is still wrong because the error
// message has been generated by cp_parser_error() which does not
// necessarily allow accurate column number display. At some point, we will
// need make cp_parser_error() report more accurate column numbers.
// { dg-error "30: error: expected '\{' at end of input" "" { target *-*-* } { 4 } }

// { dg-error "35: error: expected unqualified-id before '\}' token" "" {target *-*-*  } { 4 } }

// { dg-error "35: error: expected declaration before '\}' token" "" {target *-*-*  } { 4 } }

