// { dg-options "-fshow-column" }
// PR c++/10779

static void InstantiateConstraint(const float&, unsigned,
                                  void(*AddFunction)(const TYPE&,bool&,
                                                     char*, char*,
                                                     unsigned*));
// { dg-error "64: error: expected ',' or '...' before '&' token" "" { target *-*-* }  { 5 } }

/// in the coming test, the column information is broken as it points to
//  the end of the declaration instead of pointing to the begining of the
//  'TYPE' identifier. This is due to the warning being generated by the
//  declaration groking code (gcc/cp/decl.c) and not the parser. So in that
//  code, the exact token location information is lost as the declaration
//  groking code manipulates TREEs only. The token location used is then
//  the global one that is not accurate enough. Anyway, let's say it is
//  good enough for now, until we find a way to propagate token location to
//  code paths that manipulate TREEs only.
// { dg-error "64: error: ISO C\\+\\+ forbids declaration of 'TYPE' with no type" "" { target *-*-* } { 7 } }
